packages/sanity/src/core/store/_legacy/document/utils/sequentializeListenerEvents.ts

```diff
@@ -0,0 +1,125 @@
+/* eslint-disable no-console */
+import {concat, from, type Observable, of, switchMap, throwError, timer} from 'rxjs'
+import {mergeMap, scan} from 'rxjs/operators'
+
+import {type ListenerEventWithSnapshot} from '../getPairListener'
+import {type MutationEvent} from '../types'
+import {discardChainTo, linkedSort} from './eventChainUtils'
+
+type State = {
+  base: {revision: string | undefined} | undefined
+  events: ListenerEventWithSnapshot[] | undefined
+  buffer: MutationEvent[]
+  initialSnapshotRevision: string | undefined
+}
+
+export class OutOfSyncError extends Error {
+  constructor(message: string) {
+    super(message)
+    this.name = 'OutOfSyncError'
+  }
+}
+
+const DEFAULT_MAX_BUFFER_SIZE = 10
+const DEFAULT_DEADLINE_MS = 3000
+
+/**
+ * Takes an observable of listener events that might arrive out of order, and emits them in sequence
+ * TODO: support a threshold for when to give up
```

Comment:
is this TODO still relevant? it seems like there are thresholds for when to give up.